llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp

```diff
@@ -3597,10 +3597,22 @@ bool AArch64InstructionSelector::selectBrJT(MachineInstr &I,
   unsigned JTI = I.getOperand(1).getIndex();
   Register Index = I.getOperand(2).getReg();
 
+  MF->getInfo<AArch64FunctionInfo>()->setJumpTableEntryInfo(JTI, 4, nullptr);
+  if (MF->getFunction().hasFnAttribute("jump-table-hardening") ||
+      STI.getTargetTriple().isArm64e()) {
+    if (TM.getCodeModel() != CodeModel::Small)
```

Comment:
I might be missing smth, but why we don't support large code model here and support it in `AArch64TargetLowering::LowerBR_JT`? The offset computation is done in pseudo expansion and it's common for both ISel's, and full 64-bit offsets are supported via movk. So, we can probably support large code model here as well?

If such behavior is expected - the error message below looks untested. I'm OK with adding tests as a follow-up patch later if it's more convenient.

Reply:
Yeah, SDAG/GISel should both have the same support.  The situation with the large code model is more complicated.
Currently, the AsmPrinter expansion materializes the jump-table address using `adrp`/`add`.  On MachO, that's okay(-ish) for the large code-model as well, but that's not what's done on ELF (per `AArch64TargetLowering::LowerJumpTable`).
I don't see an explicit rationale in the code or history, but I could justify it with ELF always using a different section for the jump-table itself, vs. MachO ~always having it inline with the function.  In the latter case, I think we can usually get away without the 4x `movk` large code-model materialization sequence, so we end up with the same desired `adrp`/`add` codegen for large/small code models on MachO, but not ELF.
Of course we can teach the expansion about non-`adrp`/`add` materialization variants, but let's add that separately.

Reply:
Alternatively, I think it would be OK to explicitly not support large code model at all (both in SDAG/GISel) in context of this PR, and add such support separately to have main logic for the most common use case already merged.